### ISSUE TITLE
Parallel processing of scripts in Flutter Web

### ISSUE THREAD

**Issue by AbhishekDoshi26** [flutter/flutter#143867]
### Use case

We all know that the initial load time of Flutter web app is a bit high due to which it becomes an unlikely decision to not use Flutter Web in production. This proposal tries to reduce the initial load time.

### Proposal

I was doing some research on the same and thought of introducing parallel processing of the scripts that are present in index.html

I am not a native web developer and google search, chatGPT and Bard was at rescue. Luckily, I was able to implement parallel processing of the scripts with `defer` tag and the initial load time reduced drastically.

I have also wrote an article on the same [here](https://abhishekdoshi26.medium.com/faster-flutter-web-app-9d50eb0b6954).

I have implemented the same in my business website: https://theswagbox.in 

You can check it out in action and even try to open in incognito and disable cache, still the load time is much faster.

This needs to be implemented from the Flutter side so that all projects that are created can have the `defer` attribute in the script tag.

Give it a thought :)

**Comment by AbhishekDoshi26** [comment]
@TahaTesser wanna give it a thought?

**Comment by TahaTesser** [comment]
Interesting idea, I'm no expert in Flutter web. This looks similar to https://github.com/flutter/flutter/issues/54105

cc: @yjbanov @

**Comment by AbhishekDoshi26** [comment]
> Interesting idea, I'm no expert in Flutter web. This looks similar to #54105
> 
> cc: @yjbanov

Kind of similar, the other issue mentions for localization while I am specifically talking about loading the entire flutter web app deferred.

let’s see if anyone from the core team can check this :)

**Comment by SEGVeenstra** [comment]
Not my cup of tea, but seeing how blazing fast your personal website is, this seems like a nice win.

**Comment by AbhishekDoshi26** [comment]
> Not my cup of tea, but seeing how blazing fast your personal website is, this seems like a nice win.

Agreed! If anyone knows someone from the Flutter team or from the community who has done some research work in Flutter web, please tag them to check this proposal. Because, I feel, if this has no cons, and if this got into the SDK, then the perspective of people that Flutter Web is slow, Flutter Web is not for production, etc., will chaneg.

**Comment by adamped** [comment]
The defer tag allows the JS file to download in the background without blocking the HTML rendering. Hence since your site has a splashscreen it allows that to render faster for the user which is a good win.

Now I was about to say it doesn't seem to make much difference but testing it out in the network inspector in Chrome it does seem to start the download of other items faster. 

If you look at main.dart.js it does look like it blocks unless you mark it defer.

Without defer
![image](https://github.com/flutter/flutter/assets/13672662/bbdc0319-d0ce-470a-84f4-31366ad9c24b)

With defer
![image](https://github.com/flutter/flutter/assets/13672662/e5784262-05dc-483f-be7c-67ac69744967)

**Comment by AbhishekDoshi26** [comment]
Hmmm okk, so this solution can actually make the web app faster. Yayy!

**Comment by kevmoo** [comment]
@eyebrowsoffire – we've made some changes recently here.

**Comment by eyebrowsoffire** [comment]
I am pretty sure `defer` is not faster than `async`. The main difference between `defer` and `async` is that `defer` attempts to maintain the execution order of scripts and `async` doesn't bother with any ordering requirements.

**Comment by jeetsdev** [comment]
@eyebrowsoffire In addition to upholding the order of execution, `defer` waits for `HTML` to load before beginning the scripts; in contrast, `async` initiates script execution as soon as the scripts are downloaded. So we can say it's better to use `defer` than `async` while loading scripts.

**Comment by eyebrowsoffire** [comment]
@jeetsdev Why would that make `defer` preferable to `async`? From your description, `async` would result in less blocking/waiting, as the script wouldn't have to wait for HTML to load. FWIW you can use the new bootstrapping scripts as either `async` or `defer`.

**Comment by jeetsdev** [comment]
@eyebrowsoffire While `async` would reduce blocking and waiting, `defer` is generally preferable because most of the time we rely on HTML elements in our scripts and we need to load our whole HTML before running our script to avoid errors.

**Comment by eyebrowsoffire** [comment]
Yes, that makes sense in some cases. However, specifically for flutter's bootstrapping, it has been designed to not rely on the HTML elements in the body. If code inside your app relies on a certain state of the document before proceeding, my recommendation is to actually handle that inside your dart code. That way, the flutter app startup and engine initialization can proceed unblocked and in parallel, and if you have a hard ordering dependency, it can be expressed more directly later in the pipeline.

**Comment by jeetsdev** [comment]
I got it, I appreciate the clarification, @eyebrowsoffire. Handling all dependencies within the Dart code alone makes sense.
